src/Framework/AspNetCoreAnalyzers/src/Analyzers/Infrastructure/AmbiguousRoutePatternComparer.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Analyzers.Infrastructure.RoutePattern;
+
+namespace Microsoft.AspNetCore.Analyzers.Infrastructure;
+
+/// <summary>
+/// This route pattern comparer checks to see if two route patterns match the same URL and create ambiguous match exceptions at runtime.
+/// It doesn't check two routes exactly equal each other. For example, "/product/{id}" and "/product/{name}" aren't exactly equal but will match the same URL.
+/// </summary>
+internal sealed class AmbiguousRoutePatternComparer : IEqualityComparer<RoutePatternTree>
+{
+    public static AmbiguousRoutePatternComparer Instance { get; } = new();
+
+    public bool Equals(RoutePatternTree x, RoutePatternTree y)
+    {
+        if (x.Root.Parts.Length != y.Root.Parts.Length)
+        {
+            return false;
+        }
+
+        for (var i = 0; i < x.Root.Parts.Length; i++)
+        {
+            var xPart = x.Root.Parts[i];
+            var yPart = y.Root.Parts[i];
+
+            var equal = xPart switch
+            {
+                RoutePatternSegmentSeperatorNode _ => yPart is RoutePatternSegmentSeperatorNode,
```

Comment:
```suggestion
                RoutePatternSegmentSeparatorNode _ => yPart is RoutePatternSegmentSeparatorNode,
```

Reply:
What changed here? I can't figure out what the difference is.